# Convolutional Neural Network
May 6 2024

## Abstract
In order to study convolutional neural networks, we must familiarize ourselves with the available and state of the art tools, in this case TensorFlow. With this library we explore the CIFAR-10 dataset, with both an baseline model and various modifications, which seek to improve upon itself, be it through a change in architecture or rather the tuning of hyperparameters

## Introduction

After learning about machine and deep learning techniques, we must apply in a controlled environment to further improve our comprehension of these tools. Tensorflow, as one the industry standards, is the main tool we will be using, taking advantage of the integration of various datasets, in this case CIFAR-10. Through it, we will build a baseline model with certain specifications and modify them according to several tests, until we can see changes in the accuracy rate shown with model.fit() and model.evaluate(). 

## Methodology

- Create a baseline model 
- Modify the architecture of baseline model and document the changes in accuracy
- Modify hyperparamentes of baseline model and document the changes in accuracy 


## Experiment 1 Baseline Model
- Convolutional layer with 32 filters, 3×3 kernel size, and ReLU activation.
- MaxPooling layer with 2×2 pool size.
- Flatten layer.
- Dense layer with 128 units and ReLU activation.
- Dense output layer with softmax activation.

### Train with CIFAR-10 dataset

Dependencies

In [142]:
import tensorflow as tf
from tensorflow.keras.layers import Dense, Flatten, Conv2D, MaxPooling2D
from tensorflow.keras import Model


cifar10 = tf.keras.datasets.cifar10

Load and Prepare Data

In [143]:
(x_train, y_train), (x_test, y_test) = cifar10.load_data()
x_train, x_test = x_train / 255.0, x_test / 255.0




Create Model

In [167]:
class MyModel(Model):
  def __init__(self):
    super().__init__()
    self.conv1 = Conv2D(filters = 32, kernel_size=(3,3), activation='relu')
    self.maxpool = MaxPooling2D(pool_size=(2, 2))
    self.flatten = Flatten()
    self.d1 = Dense(units=128,activation='relu')
    self.d2 = Dense(units=10, activation="softmax")


  def call(self, x):
    x = self.conv1(x)
    x = self.flatten(x)
    x = self.d1(x)
    return self.d2(x)

# Create an instance of the model
model = MyModel()




Make Predictions

In [168]:
predictions = model(x_train[:1]).numpy()
tf.nn.softmax(predictions).numpy()
loss_fn = tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True)
loss_fn(y_train[:1], predictions).numpy()
model.compile(optimizer='adam',
              loss=loss_fn,
              metrics=['accuracy'])

c:\Users\santi\AppData\Local\Programs\Python\Python311\Lib\site-packages\keras\src\layers\layer.py:360: UserWarning: `build()` was called on layer 'my_model_45', however the layer does not have a `build()` method implemented and it looks like it has unbuilt state. This will cause the layer to be marked as built, despite not being actually built, which may cause failures down the line. Make sure to implement a proper `build()` method.
  warnings.warn(


### Fit Model

In [170]:
model.fit(x_train, y_train, epochs=5)


Epoch 1/5
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 67s 42ms/step - accuracy: 0.3568 - loss: 1.8377
Epoch 2/5
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 69s 44ms/step - accuracy: 0.5498 - loss: 1.2786
Epoch 3/5
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 75s 48ms/step - accuracy: 0.5948 - loss: 1.1488
Epoch 4/5
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 78s 50ms/step - accuracy: 0.6254 - loss: 1.0633
Epoch 5/5
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 86s 55ms/step - accuracy: 0.6578 - loss: 0.9706


In [171]:
model.evaluate(x_test,  y_test, verbose=2)


313/313 - 1s - 3ms/step - accuracy: 0.5898 - loss: 1.1744


[1.1743977069854736, 0.5898000001907349]

### Recorded accuracy: 0.5898

## Experiment 2 Changing Architecture 

#### Changing filters to 48

In [147]:
class MyModel(Model):
  def __init__(self):
    super().__init__()
    self.conv1 = Conv2D(filters = 48, kernel_size=(3,3), activation='relu')
    self.maxpool = MaxPooling2D(pool_size=(2, 2))
    self.flatten = Flatten()
    self.d1 = Dense(units=128,activation='relu')
    self.d2 = Dense(units=10, activation="softmax")


  def call(self, x):
    x = self.conv1(x)
    x = self.flatten(x)
    x = self.d1(x)
    return self.d2(x)

# Create an instance of the model
model = MyModel()

In [148]:
predictions = model(x_train[:1]).numpy()
tf.nn.softmax(predictions).numpy()
loss_fn = tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True)
loss_fn(y_train[:1], predictions).numpy()
model.compile(optimizer='adam',
              loss=loss_fn,
              metrics=['accuracy'])

c:\Users\santi\AppData\Local\Programs\Python\Python311\Lib\site-packages\keras\src\layers\layer.py:360: UserWarning: `build()` was called on layer 'my_model_40', however the layer does not have a `build()` method implemented and it looks like it has unbuilt state. This will cause the layer to be marked as built, despite not being actually built, which may cause failures down the line. Make sure to implement a proper `build()` method.
  warnings.warn(


In [150]:
model.fit(x_train, y_train, epochs=5)


Epoch 1/5
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 88s 56ms/step - accuracy: 0.4080 - loss: 1.6475
Epoch 2/5
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 87s 56ms/step - accuracy: 0.5759 - loss: 1.2005
Epoch 3/5
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 86s 55ms/step - accuracy: 0.6387 - loss: 1.0281
Epoch 4/5
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 87s 55ms/step - accuracy: 0.6896 - loss: 0.8927
Epoch 5/5
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 94s 60ms/step - accuracy: 0.7380 - loss: 0.7548


In [151]:
model.evaluate(x_test,  y_test, verbose=2)

313/313 - 2s - 5ms/step - accuracy: 0.5980 - loss: 1.2176


[1.217604160308838, 0.5979999899864197]

### Recorded Accuracy: 0.5980
0.0082 Increase

#### Duplicating the convolutional layer

In [153]:
class MyModel(Model):
  def __init__(self):
    super().__init__()
    self.conv1 = Conv2D(filters = 48, kernel_size=(3,3), activation='relu')
    self.conv2 = Conv2D(filters = 48, kernel_size=(3,3), activation='relu')
    self.maxpool = MaxPooling2D(pool_size=(2, 2))
    self.flatten = Flatten()
    self.d1 = Dense(units=128,activation='relu')
    self.d2 = Dense(units=10, activation="softmax")


  def call(self, x):
    x = self.conv1(x)
    x = self.conv2(x)
    x = self.flatten(x)
    x = self.d1(x)
    return self.d2(x)

# Create an instance of the model
model = MyModel()

In [154]:
predictions = model(x_train[:1]).numpy()
tf.nn.softmax(predictions).numpy()
loss_fn = tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True)
loss_fn(y_train[:1], predictions).numpy()
model.compile(optimizer='adam',
              loss=loss_fn,
              metrics=['accuracy'])

c:\Users\santi\AppData\Local\Programs\Python\Python311\Lib\site-packages\keras\src\layers\layer.py:360: UserWarning: `build()` was called on layer 'my_model_42', however the layer does not have a `build()` method implemented and it looks like it has unbuilt state. This will cause the layer to be marked as built, despite not being actually built, which may cause failures down the line. Make sure to implement a proper `build()` method.
  warnings.warn(


In [155]:
model.fit(x_train, y_train, epochs=5)

Epoch 1/5


c:\Users\santi\AppData\Local\Programs\Python\Python311\Lib\site-packages\keras\src\backend\tensorflow\nn.py:625: UserWarning: "`sparse_categorical_crossentropy` received `from_logits=True`, but the `output` argument was produced by a Softmax activation and thus does not represent logits. Was this intended?
  output, from_logits = _get_logits(


1563/1563 ━━━━━━━━━━━━━━━━━━━━ 88s 56ms/step - accuracy: 0.4470 - loss: 1.5470
Epoch 2/5
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 92s 59ms/step - accuracy: 0.6736 - loss: 0.9315
Epoch 3/5
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 133s 85ms/step - accuracy: 0.7613 - loss: 0.6846
Epoch 4/5
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 129s 82ms/step - accuracy: 0.8413 - loss: 0.4613
Epoch 5/5
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 137s 87ms/step - accuracy: 0.9116 - loss: 0.2671


In [156]:
model.evaluate(x_test,  y_test, verbose=2)

313/313 - 2s - 7ms/step - accuracy: 0.6334 - loss: 1.4311


[1.4311282634735107, 0.633400022983551]

### Recorded Accuracy: 0.6334
0.0436 Increase

#### Adding more convolutional and dense layers, modifying filters and units

In [157]:
class MyModel(Model):
  def __init__(self):
    super().__init__()
    self.conv1 = Conv2D(filters = 32, kernel_size=(3,3), activation='relu')
    self.conv2 = Conv2D(filters = 24, kernel_size=(3,3), activation='relu')
    self.conv3 = Conv2D(filters = 12, kernel_size=(3,3), activation='relu')
    self.maxpool = MaxPooling2D(pool_size=(2, 2))
    self.flatten = Flatten()
    self.d1 = Dense(units=128,activation='relu')
    self.d3 = Dense(units=64,activation='relu')
    self.d2 = Dense(units=10, activation="softmax")


  def call(self, x):
    x = self.conv1(x)
    x = self.conv2(x)
    x = self.conv3(x)
    x = self.flatten(x)
    x = self.d1(x)
    x = self.d3(x)
    return self.d2(x)

# Create an instance of the model
model = MyModel()

In [158]:
predictions = model(x_train[:1]).numpy()
tf.nn.softmax(predictions).numpy()
loss_fn = tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True)
loss_fn(y_train[:1], predictions).numpy()
model.compile(optimizer='adam',
              loss=loss_fn,
              metrics=['accuracy'])

c:\Users\santi\AppData\Local\Programs\Python\Python311\Lib\site-packages\keras\src\layers\layer.py:360: UserWarning: `build()` was called on layer 'my_model_43', however the layer does not have a `build()` method implemented and it looks like it has unbuilt state. This will cause the layer to be marked as built, despite not being actually built, which may cause failures down the line. Make sure to implement a proper `build()` method.
  warnings.warn(


In [159]:
model.fit(x_train, y_train, epochs=5)

Epoch 1/5


c:\Users\santi\AppData\Local\Programs\Python\Python311\Lib\site-packages\keras\src\backend\tensorflow\nn.py:625: UserWarning: "`sparse_categorical_crossentropy` received `from_logits=True`, but the `output` argument was produced by a Softmax activation and thus does not represent logits. Was this intended?
  output, from_logits = _get_logits(


1563/1563 ━━━━━━━━━━━━━━━━━━━━ 26s 16ms/step - accuracy: 0.3716 - loss: 1.7084
Epoch 2/5
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 25s 16ms/step - accuracy: 0.5944 - loss: 1.1446
Epoch 3/5
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 24s 15ms/step - accuracy: 0.6571 - loss: 0.9677
Epoch 4/5
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 28s 18ms/step - accuracy: 0.7097 - loss: 0.8227
Epoch 5/5
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 22s 14ms/step - accuracy: 0.7569 - loss: 0.6931


In [160]:
model.evaluate(x_test,  y_test, verbose=2)

313/313 - 1s - 4ms/step - accuracy: 0.6507 - loss: 1.0408


[1.0408152341842651, 0.6506999731063843]

### Recorded Accuracy: 0.6507
0.0609 Increase

In [177]:
class MyModel(Model):
  def __init__(self):
    super().__init__()
    self.conv1 = Conv2D(filters = 32, kernel_size=(3,3), activation='relu')
    self.conv2 = Conv2D(filters = 32, kernel_size=(3,3), activation='relu')
    self.conv3 = Conv2D(filters = 32, kernel_size=(3,3), activation='relu')
    self.conv4 = Conv2D(filters = 32, kernel_size=(3,3), activation='relu')
    self.maxpool = MaxPooling2D(pool_size=(2, 2))
    self.flatten = Flatten()
    self.d1 = Dense(units=128,activation='relu')
    self.d3 = Dense(units=128,activation='relu')
    self.d2 = Dense(units=10, activation="softmax")


  def call(self, x):
    x = self.conv1(x)
    x = self.conv2(x)
    x = self.conv3(x)
    x = self.conv4(x)
    x = self.flatten(x)
    x = self.d1(x)
    x = self.d3(x)
    return self.d2(x)

# Create an instance of the model
model = MyModel()

In [178]:
predictions = model(x_train[:1]).numpy()
tf.nn.softmax(predictions).numpy()
loss_fn = tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True)
loss_fn(y_train[:1], predictions).numpy()
model.compile(optimizer='adam',
              loss=loss_fn,
              metrics=['accuracy'])

c:\Users\santi\AppData\Local\Programs\Python\Python311\Lib\site-packages\keras\src\layers\layer.py:360: UserWarning: `build()` was called on layer 'my_model_48', however the layer does not have a `build()` method implemented and it looks like it has unbuilt state. This will cause the layer to be marked as built, despite not being actually built, which may cause failures down the line. Make sure to implement a proper `build()` method.
  warnings.warn(


In [179]:
model.fit(x_train, y_train, epochs=5)

Epoch 1/5


c:\Users\santi\AppData\Local\Programs\Python\Python311\Lib\site-packages\keras\src\backend\tensorflow\nn.py:625: UserWarning: "`sparse_categorical_crossentropy` received `from_logits=True`, but the `output` argument was produced by a Softmax activation and thus does not represent logits. Was this intended?
  output, from_logits = _get_logits(


1563/1563 ━━━━━━━━━━━━━━━━━━━━ 53s 33ms/step - accuracy: 0.3885 - loss: 1.6664
Epoch 2/5
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 51s 33ms/step - accuracy: 0.6193 - loss: 1.0694
Epoch 3/5
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 51s 33ms/step - accuracy: 0.7089 - loss: 0.8179
Epoch 4/5
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 52s 33ms/step - accuracy: 0.7850 - loss: 0.6214
Epoch 5/5
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 52s 33ms/step - accuracy: 0.8534 - loss: 0.4178


In [180]:
model.evaluate(x_test,  y_test, verbose=2)

313/313 - 2s - 6ms/step - accuracy: 0.6607 - loss: 1.1505


[1.150483250617981, 0.6607000231742859]

### Recorded Accuracy: 0.6607
0.0709 Increase

## Experiment 3 Hyperparameter Tuning

### Increase Batch

In [181]:
class MyModel(Model):
  def __init__(self):
    super().__init__()
    self.conv1 = Conv2D(filters = 32, kernel_size=(3,3), activation='relu')
    self.maxpool = MaxPooling2D(pool_size=(2, 2))
    self.flatten = Flatten()
    self.d1 = Dense(units=128,activation='relu')
    self.d2 = Dense(units=128, activation="softmax")


  def call(self, x):
    x = self.conv1(x)
    x = self.flatten(x)
    x = self.d1(x)
    return self.d2(x)

# Create an instance of the model
model = MyModel()

In [182]:
predictions = model(x_train[:1]).numpy()
tf.nn.softmax(predictions).numpy()
loss_fn = tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True)
loss_fn(y_train[:1], predictions).numpy()
model.compile(optimizer='adam',
              loss=loss_fn,
              metrics=['accuracy'])

c:\Users\santi\AppData\Local\Programs\Python\Python311\Lib\site-packages\keras\src\layers\layer.py:360: UserWarning: `build()` was called on layer 'my_model_49', however the layer does not have a `build()` method implemented and it looks like it has unbuilt state. This will cause the layer to be marked as built, despite not being actually built, which may cause failures down the line. Make sure to implement a proper `build()` method.
  warnings.warn(


In [183]:
model.fit(x_train, y_train, epochs=5)

Epoch 1/5


c:\Users\santi\AppData\Local\Programs\Python\Python311\Lib\site-packages\keras\src\backend\tensorflow\nn.py:625: UserWarning: "`sparse_categorical_crossentropy` received `from_logits=True`, but the `output` argument was produced by a Softmax activation and thus does not represent logits. Was this intended?
  output, from_logits = _get_logits(


1563/1563 ━━━━━━━━━━━━━━━━━━━━ 63s 40ms/step - accuracy: 0.3995 - loss: 1.7264
Epoch 2/5
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 59s 38ms/step - accuracy: 0.5789 - loss: 1.1998
Epoch 3/5
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 61s 39ms/step - accuracy: 0.6433 - loss: 1.0277
Epoch 4/5
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 88s 56ms/step - accuracy: 0.6945 - loss: 0.8688
Epoch 5/5
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 104s 67ms/step - accuracy: 0.7449 - loss: 0.7350


In [184]:
model.evaluate(x_test,  y_test, verbose=2)

313/313 - 1s - 4ms/step - accuracy: 0.6155 - loss: 1.1896


[1.189598560333252, 0.6154999732971191]

### Recorded Accuracy: 0.6155
0.0257 Increase

### Increase Epochs

In [185]:
class MyModel(Model):
  def __init__(self):
    super().__init__()
    self.conv1 = Conv2D(filters = 32, kernel_size=(3,3), activation='relu')
    self.maxpool = MaxPooling2D(pool_size=(2, 2))
    self.flatten = Flatten()
    self.d1 = Dense(units=128,activation='relu')
    self.d2 = Dense(units=10, activation="softmax")


  def call(self, x):
    x = self.conv1(x)
    x = self.flatten(x)
    x = self.d1(x)
    return self.d2(x)

# Create an instance of the model
model = MyModel()

In [186]:
predictions = model(x_train[:1]).numpy()
tf.nn.softmax(predictions).numpy()
loss_fn = tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True)
loss_fn(y_train[:1], predictions).numpy()
model.compile(optimizer='adam',
              loss=loss_fn,
              metrics=['accuracy'])

c:\Users\santi\AppData\Local\Programs\Python\Python311\Lib\site-packages\keras\src\layers\layer.py:360: UserWarning: `build()` was called on layer 'my_model_50', however the layer does not have a `build()` method implemented and it looks like it has unbuilt state. This will cause the layer to be marked as built, despite not being actually built, which may cause failures down the line. Make sure to implement a proper `build()` method.
  warnings.warn(


In [187]:
model.fit(x_train, y_train, epochs=10)

Epoch 1/10
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 87s 54ms/step - accuracy: 0.4151 - loss: 1.6407
Epoch 2/10
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 81s 52ms/step - accuracy: 0.6073 - loss: 1.1233
Epoch 3/10
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 86s 55ms/step - accuracy: 0.6790 - loss: 0.9175
Epoch 4/10
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 86s 55ms/step - accuracy: 0.7314 - loss: 0.7701
Epoch 5/10
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 81s 52ms/step - accuracy: 0.7811 - loss: 0.6316
Epoch 6/10
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 80s 51ms/step - accuracy: 0.8308 - loss: 0.5006
Epoch 7/10
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 81s 52ms/step - accuracy: 0.8703 - loss: 0.3849
Epoch 8/10
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 84s 54ms/step - accuracy: 0.9037 - loss: 0.2919
Epoch 9/10
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 86s 55ms/step - accuracy: 0.9228 - loss: 0.2306
Epoch 10/10
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 69s 44ms/step - accuracy: 0.9466 - loss: 0.1670


In [189]:
model.evaluate(x_test,  y_test, verbose=2)

313/313 - 1s - 3ms/step - accuracy: 0.5778 - loss: 2.0718


[2.0717928409576416, 0.5777999758720398]

### Recorded Accuracy: 0.5778
0.012 Decrease

### Combinations

In [190]:
class MyModel(Model):
  def __init__(self):
    super().__init__()
    self.conv1 = Conv2D(filters = 32, kernel_size=(3,3), activation='relu')
    self.maxpool = MaxPooling2D(pool_size=(2, 2))
    self.flatten = Flatten()
    self.d1 = Dense(units=128,activation='relu')
    self.d2 = Dense(units=128, activation="softmax")


  def call(self, x):
    x = self.conv1(x)
    x = self.flatten(x)
    x = self.d1(x)
    return self.d2(x)

# Create an instance of the model
model = MyModel()

In [191]:
predictions = model(x_train[:1]).numpy()
tf.nn.softmax(predictions).numpy()
loss_fn = tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True)
loss_fn(y_train[:1], predictions).numpy()
model.compile(optimizer='adam',
              loss=loss_fn,
              metrics=['accuracy'])

c:\Users\santi\AppData\Local\Programs\Python\Python311\Lib\site-packages\keras\src\layers\layer.py:360: UserWarning: `build()` was called on layer 'my_model_51', however the layer does not have a `build()` method implemented and it looks like it has unbuilt state. This will cause the layer to be marked as built, despite not being actually built, which may cause failures down the line. Make sure to implement a proper `build()` method.
  warnings.warn(


In [192]:
model.fit(x_train, y_train, epochs=7)

Epoch 1/7
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 58s 36ms/step - accuracy: 0.4179 - loss: 1.6830
Epoch 2/7
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 57s 36ms/step - accuracy: 0.5946 - loss: 1.1531
Epoch 3/7
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 58s 37ms/step - accuracy: 0.6666 - loss: 0.9567
Epoch 4/7
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 58s 37ms/step - accuracy: 0.7215 - loss: 0.8015
Epoch 5/7
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 59s 38ms/step - accuracy: 0.7875 - loss: 0.6260
Epoch 6/7
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 58s 37ms/step - accuracy: 0.8384 - loss: 0.4812
Epoch 7/7
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 58s 37ms/step - accuracy: 0.8804 - loss: 0.3538


In [193]:
model.evaluate(x_test,  y_test, verbose=2)

313/313 - 1s - 3ms/step - accuracy: 0.6104 - loss: 1.4845


[1.4845486879348755, 0.6104000210762024]

In [203]:
class MyModel(Model):
  def __init__(self):
    super().__init__()
    self.conv1 = Conv2D(filters = 32, kernel_size=(3,3), activation='relu')
    self.maxpool = MaxPooling2D(pool_size=(3, 3))
    self.flatten = Flatten()
    self.d1 = Dense(units=256,activation='relu')
    self.d2 = Dense(units=256, activation="softmax")


  def call(self, x):
    x = self.conv1(x)
    x = self.flatten(x)
    x = self.d1(x)
    return self.d2(x)

# Create an instance of the model
model = MyModel()

In [204]:
predictions = model(x_train[:1]).numpy()
tf.nn.softmax(predictions).numpy()
loss_fn = tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True)
loss_fn(y_train[:1], predictions).numpy()
model.compile(optimizer='adam',
              loss=loss_fn,
              metrics=['accuracy'])

c:\Users\santi\AppData\Local\Programs\Python\Python311\Lib\site-packages\keras\src\layers\layer.py:360: UserWarning: `build()` was called on layer 'my_model_54', however the layer does not have a `build()` method implemented and it looks like it has unbuilt state. This will cause the layer to be marked as built, despite not being actually built, which may cause failures down the line. Make sure to implement a proper `build()` method.
  warnings.warn(


In [205]:
model.fit(x_train, y_train, epochs=6)

Epoch 1/6
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 113s 72ms/step - accuracy: 0.4251 - loss: 1.6663
Epoch 2/6
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 112s 72ms/step - accuracy: 0.6169 - loss: 1.0905
Epoch 3/6
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 112s 72ms/step - accuracy: 0.6949 - loss: 0.8667
Epoch 4/6
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 115s 74ms/step - accuracy: 0.7702 - loss: 0.6644
Epoch 5/6
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 114s 73ms/step - accuracy: 0.8355 - loss: 0.4838
Epoch 6/6
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 113s 72ms/step - accuracy: 0.8934 - loss: 0.3194


In [207]:
model.evaluate(x_test,  y_test, verbose=2)

313/313 - 1s - 4ms/step - accuracy: 0.6131 - loss: 1.4894


[1.4894487857818604, 0.613099992275238]

### Recorded Accuracy: 0.6131
0.0233 Increase

## Results and Discussion
After observing the effects of both architecture and hyperparamenters, we can see both affect the resulting accuracy rate of each model. Though we could see greater variance with architectural changes, both seem to be independent and should help create better models through careful tuning. 

It is hard to say what each component affects what at a glance, but there is a clear relation with execution time and complexity of the model, which in turn is not directly correlated with a better solution. Rather, overly complex models gave poorer results due to overfitting, observed by the model.evaluate() function. 

## Conclusion
Deep learning tools such as Tensorflow provide with a quick jumpstart into machine learning, making it easy to directly interact with the model through different parameters and making it an important tool to further study such algorithms and apply them to real life scenarios. Though further learning is necesary to discover more functionalities and gain a sense of the correct approach to different datasets, the usage of the framework presents many opportunities to do so. 

## References
Tensorflow (2024) Tensorflow Core: Tutorials https://www.tensorflow.org/tutorials?hl=en